// multi-proc@1.1.0/lib/NodeProcess.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = require("child_process");
const path_1 = require("path");
const Actions_1 = require("./Actions");
const Colors_1 = require("./Colors");
class NodeProcess {
    constructor(inputProc, tagWidth, filter) {
        this.log = (message) => {
            const lines = message.trim().split('\n');
            for (const line of lines) {
                console.log(`${this.tag}${line.trim()}`);
            }
        };
        this.error = (message) => {
            const lines = message.trim().split('\n');
            for (const line of lines) {
                console.log(`${Colors_1.color(this.tag, Colors_1.Color.BG_RED, Colors_1.Color.UNDERLINE, Colors_1.Color.WHITE)}${line.trim()}`);
            }
        };
        this.getTag = (proc, maxWidth) => {
            let tag = proc.tag;
            while (tag.length < maxWidth) {
                tag += ' ';
            }
            if (!proc.color) {
                throw new Error(`'process.color' is required`);
            }
            // @ts-ignore
            const color = Colors_1.Color[proc.color.toUpperCase()];
            if (!color) {
                throw new Error(`${proc.color.toUpperCase()} is not a valid color.`);
            }
            return Colors_1.color(` ${tag} `, color, Colors_1.Color.WHITE) + ' ';
        };
        this.tag = this.getTag(inputProc, tagWidth);
        this.command = inputProc.command;
        this.filter = filter;
        if (inputProc.directory) {
            if (inputProc.directory.startsWith('/')) {
                this.directory = inputProc.directory;
            }
            else {
                this.directory = path_1.join(Actions_1.default.pwd, inputProc.directory);
            }
        }
    }
}
NodeProcess.start = (proc) => __awaiter(this, void 0, void 0, function* () {
    return new Promise((resolve, reject) => {
        const spawnOptions = {
            cwd: proc.directory,
            shell: proc.command.search(/(&&|;)/) >= 0,
        };
        const startMessageArray = [`${Colors_1.color('Starting: ', Colors_1.Color.BOLD)}${proc.command}`];
        if (spawnOptions.cwd) {
            startMessageArray.push(Colors_1.color(`in ${spawnOptions.cwd}`, Colors_1.Color.DIM));
        }
        if (spawnOptions.shell) {
            startMessageArray.push(Colors_1.color(`(as a shell)`, Colors_1.Color.DIM));
        }
        proc.log(startMessageArray.join(' '));
        let spawned;
        if (spawnOptions.shell) {
            spawned = child_process_1.spawn(proc.command, spawnOptions);
        }
        else {
            const commandArray = proc.command.split(' ');
            const initialCommand = commandArray.shift();
            spawned = child_process_1.spawn(initialCommand, commandArray, spawnOptions);
        }
        spawned.addListener('close', () => resolve());
        spawned.addListener('disconnect', () => resolve());
        spawned.addListener('exit', () => {
            proc.log(Colors_1.color('Finished!', Colors_1.Color.BOLD));
            resolve();
        });
        spawned.addListener('err', (err) => reject(err));
        spawned.stdout.addListener('data', (chunk) => {
            if (!proc.filter || proc.filter.exec(proc.tag)) {
                proc.log(chunk.toString());
            }
        });
        spawned.stderr.addListener('data', (chunk) => {
            if (!proc.filter || proc.filter.exec(proc.tag)) {
                proc.log(chunk.toString());
            }
        });
    });
});
exports.default = NodeProcess;
